From cd1fcc073c46748a8f32746f262d63432fed22c9 Mon Sep 17 00:00:00 2001
From: Stefan Assmann <sassmann@redhat.com>
Date: Tue, 2 Sep 2014 11:37:31 -0400
Subject: [ethernet] i40e: Retain MAC filters when changing port VLAN

Message-id: <1409658042-22689-91-git-send-email-sassmann@redhat.com>
Patchwork-id: 90285
O-Subject: [RHEL7.1 PATCH 090/281] i40e: Retain MAC filters when changing port VLAN
Bugzilla: 1091126
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Greg Rose <gregory.v.rose@intel.com>

We were already taking care to retain the MAC filter list when deleting a
port VLAN.  Take some additional care to retain the MAC filter list when
changing a port VLAN.

Change-ID: Iacf9599ea24ecb4dca8e419aacaf4b58ca361a9c
Signed-off-by: Greg Rose <gregory.v.rose@intel.com>
Signed-off-by: Catherine Sullivan <catherine.sullivan@intel.com>
Tested-by: Sibai Li <sibai.li@intel.com>
Signed-off-by: Jeff Kirsher <jeffrey.t.kirsher@intel.com>
Signed-off-by: Stefan Assmann <sassmann@redhat.com>
(cherry picked from commit 1315f7c34e08f50b0d710168c2d98fb4766a7f5d)
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
index b27b2f5..982eef0 100644
--- a/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
+++ b/drivers/net/ethernet/intel/i40e/i40e_virtchnl_pf.c
@@ -2128,11 +2128,15 @@ int i40e_ndo_set_vf_port_vlan(struct net_device *netdev,
 
  /* Check for condition where there was already a port VLAN ID
   * filter set and now it is being deleted by setting it to zero.
+  * Additionally check for the condition where there was a port
+  * VLAN but now there is a new and different port VLAN being set.
   * Before deleting all the old VLAN filters we must add new ones
   * with -1 (I40E_VLAN_ANY) or otherwise we're left with all our
   * MAC addresses deleted.
   */
- if (!(vlan_id || qos) && vsi->info.pvid)
+ if ((!(vlan_id || qos) ||
+     (vlan_id | qos) != le16_to_cpu(vsi->info.pvid)) &&
+     vsi->info.pvid)
   ret = i40e_vsi_add_vlan(vsi, I40E_VLAN_ANY);
 
  if (vsi->info.pvid) {
-- 
1.7.1